src/runtime-rs/crates/runtimes/virt_container/src/health_check.rs

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) 2019-2022 Alibaba Cloud
+// Copyright (c) 2019-2022 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::sync::Arc;
+
+use agent::Agent;
+use anyhow::Context;
+use tokio::sync::{mpsc, Mutex};
+
+/// monitor check interval 30s
+const HEALTH_CHECK_TIMER_INTERVAL: u64 = 30;
+
+/// health check threshold 5min
+const HEALTH_CHECK_THRESHOLD: u64 = 5 * 60 / HEALTH_CHECK_TIMER_INTERVAL;
+
+/// health check stop channel buffer size
+const HEALTH_CHECK_STOP_CHANNEL_BUFFER_SIZE: usize = 1;
+
+pub struct HealthCheck {
+    pub keep_alive: bool,
+    keep_vm: bool,
+    stop_tx: mpsc::Sender<()>,
+    stop_rx: Arc<Mutex<mpsc::Receiver<()>>>,
+}
+
+impl HealthCheck {
+    pub fn new(keep_alive: bool, keep_vm: bool) -> HealthCheck {
+        let (tx, rx) = mpsc::channel(HEALTH_CHECK_STOP_CHANNEL_BUFFER_SIZE);
+        HealthCheck {
+            keep_alive,
+            keep_vm,
+            stop_tx: tx,
+            stop_rx: Arc::new(Mutex::new(rx)),
+        }
+    }
+
+    pub fn start(&self, id: &str, agent: Arc<dyn Agent>) {
+        if !self.keep_alive {
+            return;
+        }
+        let id = id.to_string();
+
+        info!(sl!(), "start runtime keep alive");
+
+        let stop_rx = self.stop_rx.clone();
+        let keep_vm = self.keep_vm;
+        let _ = tokio::spawn(async move {
+            let mut health_check_threshold_count = 0;
+
+            loop {
+                tokio::time::sleep(std::time::Duration::from_secs(HEALTH_CHECK_TIMER_INTERVAL))
+                    .await;
+                let mut stop_rx = stop_rx.lock().await;
+                match stop_rx.try_recv() {
+                    Ok(_) => {
+                        info!(sl!(), "revive stop {} monitor signal", id);
+                        break;
+                    }
+
+                    Err(mpsc::error::TryRecvError::Empty) => {
+                        // check agent
+                        match agent
+                            .check(agent::CheckRequest::new(""))
+                            .await
+                            .context("check health")
+                        {
+                            Ok(_) => {
+                                debug!(sl!(), "check {} agent health successfully", id);
+                                health_check_threshold_count += 1;
```

Comment:
Why check the version every 5 minutes?

Reply:
Thanks for your review, this is an issue for refactoring, the health check is in the 30s and the version check is 5min, has been updated.